### ISSUE TITLE
self:update failed. Win10, Python 3.7.2 Dead poetry.

### ISSUE THREAD

**Issue by ST33LDI9ITAL** [python-poetry/poetry#977]
- [ x] I am on the [latest](https://github.com/sdispater/poetry/releases/latest) Poetry version.
- [x ] I have searched the [issues](https://github.com/sdispater/poetry/issues) of this repo and believe that this is not a duplicate.
- [ x] If an exception occurs when executing a command, I executed it again in debug mode (`-vvv` option).

- **OS version and name**: 
- **Poetry version**: 
- **Link of a [Gist](https://gist.github.com/) with the contents of your pyproject.toml file**: 

## Issue

Went to self:update from 0.12.11 to 1.0.0.a2 preview and got this. 

c:\Python\WinPy\python-3.7.2.amd64>poetry show --version
Traceback (most recent call last):
  File "C:\Users\ST33L\.poetry\bin\poetry", line 11, in <module>
    from poetry.console import main
ModuleNotFoundError: No module named 'poetry.console'

c:\Python\WinPy\python-3.7.2.amd64>poetry self:update --preview
Traceback (most recent call last):
  File "C:\Users\ST33L\.poetry\bin\poetry", line 11, in <module>
    from poetry.console import main
ModuleNotFoundError: No module named 'poetry.console'

c:\Python\WinPy\python-3.7.2.amd64>curl -sSL https://raw.githubusercontent.com/sdispater/poetry/master/get-poetry.py | py
Retrieving Poetry metadata

Latest version already installed.

**Comment by drunkwcodes** [comment]
The error msg of updating is missing.
How to reproduce it?

What happened after re-downloading?

**Comment by ST33LDI9ITAL** [comment]
Ok, so I'm on Win10, and have multiple versions of Python, and I do things the windows way. Which is by using py launcher. So, in order for poetry to run I change poetry.bat python command to py. When updating it redownload then ofcourse it wouldn't run because of the bat changed back to python instead of py. So, I changed it back to py then reran and error still persisted. After deleting poetry folders, setting path to python, it then installed cleanly. 

Basically, doesn't support py launcher.

**Comment by drunkwcodes** [comment]
@ST33LDI9ITAL  Let me clarify that: you solved it by cleaning up manually and fixing the executable path.

The history in your cmd is worth of reserving before figuring the issue out.

I'm still wondering the debug msg of self:update command. Will you share it anyway?

**Comment by ST33LDI9ITAL** [comment]
Here ya go:

`C:\Python\poetry>poetry self:update --preview -vvv
Updating to 1.0.0a2

[PermissionError]
[WinError 5] Access is denied: 'C:\\Users\\ST33L\\.poetry\\lib\\poetry\\_vendor\\py3.7\\msgpack\\_cmsgpack.cp37-win_amd64.pyd'

Exception trace:
 C:\Users\ST33L\.poetry\lib\poetry\_vendor\py3.7\cleo\application.py in run() at line 94

 C:\Users\ST33L\.poetry\lib\poetry\console\application.py in do_run() at line 88

 C:\Users\ST33L\.poetry\lib\poetry\_vendor\py3.7\cleo\application.py in do_run() at line 197

 C:\Users\ST33L\.poetry\lib\poetry\console\commands\command.py in run() at line 77

 C:\Users\ST33L\.poetry\lib\poetry\_vendor\py3.7\cleo\commands\base_command.py in run() at line 146

 C:\Users\ST33L\.poetry\lib\poetry\_vendor\py3.7\cleo\commands\command.py in execute() at line 107

 C:\Users\ST33L\.poetry\lib\poetry\console\commands\self\update.py in handle() at line 108

 C:\Users\ST33L\.poetry\lib\poetry\console\commands\self\update.py in update() at line 132

 C:\Python\WinPy\python-3.7.2.amd64\lib\shutil.py in rmtree() at line 513
   return _rmtree_unsafe(path, onerror)
 C:\Python\WinPy\python-3.7.2.amd64\lib\shutil.py in _rmtree_unsafe() at line 392
   _rmtree_unsafe(fullname, onerror)
 C:\Python\WinPy\python-3.7.2.amd64\lib\shutil.py in _rmtree_unsafe() at line 392
   _rmtree_unsafe(fullname, onerror)
 C:\Python\WinPy\python-3.7.2.amd64\lib\shutil.py in _rmtree_unsafe() at line 392
   _rmtree_unsafe(fullname, onerror)
 C:\Python\WinPy\python-3.7.2.amd64\lib\shutil.py in _rmtree_unsafe() at line 392
   _rmtree_unsafe(fullname, onerror)
 C:\Python\WinPy\python-3.7.2.amd64\lib\shutil.py in _rmtree_unsafe() at line 397
   onerror(os.unlink, fullname, sys.exc_info())
 C:\Python\WinPy\python-3.7.2.amd64\lib\shutil.py in _rmtree_unsafe() at line 395
   os.unlink(fullname)

self:update [--preview] [--] [<version>]`

**Comment by drunkwcodes** [comment]
Is it a file permission [problem?](https://support.microsoft.com/en-gb/help/2623670/access-denied-or-other-errors-when-you-access-or-work-with-files-and-f)

**Comment by ST33LDI9ITAL** [comment]
Appears so, but only with itself, the update process. Ran under Administrator cmd. Can delete folder with no locking issues.

**Comment by ST33LDI9ITAL** [comment]
As for the rest of that.. that’s not how I set up python and not about to change. It’s my belief that if you are designing an external supplement for it that you should cover all bases. I’m simply reporting the issue.

**Comment by dilzeem** [comment]
Also getting exactly the same issue using windows 10.

I managed to solve by deleting the offending file:

`'C:\%USERPROFILE%\.poetry\lib\poetry\_vendor\py3.7\msgpack\_cmsgpack.cp37-win_amd64.pyd'`

I then ran 
`poetry self:update --preview`

**Comment by alex-lechner** [comment]
@dilzeem if I do the steps like you did (1. Deleteing the `_cmsgpack.cp37-win_amd64.pyd` and 2. execute `poetry self:update --preview` afterwards) I get the following error:

```sh
Updating to 1.0.0a3
                                                                                
[OSError]
[WinError 145] The directory is not empty: 'C:\\Users\\Alexa\\.poetry\\l
ib\\poetry\\_vendor'
                                                                                
self:update [--preview] [--] [<version>]
```

Why does installing stuff for Python on Windows has to be such a burden every single time? 😩

**Comment by dilzeem** [comment]
@alex-lechner Yeah I get the same error, I have now resorted to not using the preview version. 


Unless you have a fix???

**Comment by dilzeem** [comment]
@alex-lechner 

I ran into this same issue again, and I was able to figure it out. You need to close the windows explorer window after you have deleted the `_cmsgpack.cp37-win_amd64.pyd`.

**Comment by KelSolaar** [comment]
Just had the same problem on Github Actions: https://github.com/KelSolaar/colour/runs/238653725#step:5:52

**Comment by Sh4rK** [comment]
I ran into the same problem today.

```
C:\redacted>poetry --version
Poetry 0.12.17


C:\redacted>poetry self:update --preview
Updating to 1.0.0b2

[PermissionError]
[WinError 5] Access is denied: 'C:\\Users\\redacted\\.poetry\\lib\\poetry\\_vendor\\py3.7\\msgpack\\_cmsgpack.cp37-win_amd64.pyd'

self:update [--preview] [--] [<version>]


C:\redacted>poetry self:update --preview
Traceback (most recent call last):
  File "C:\Users\eszaant\.poetry\bin\poetry", line 12, in <module>
    from poetry.console import main
ModuleNotFoundError: No module named 'poetry.console'
```

I suspect the problem is that Windows locks currently executing files and loaded DLLs (and `pyd` files are DLLs), so replacing or deleting this file fails (though from the error message it's not clear what file operation is taking place).

But, I haven't looked at the code, I may be totally wrong.

**Comment by absassi** [comment]
I also have the same problem, and the best solution I found for me is to use [`pipx`](https://github.com/pipxproject/pipx) instead to manage Poetry installation on Windows, as [suggested in the docs](https://python-poetry.org/docs/#installing-with-pipx). It works like a charm, doesn't install its dependencies globally as it would be using `pip` and upgrading Poetry is as easy as `pipx upgrade poetry`, no locking issues.

**Comment by nicolaerario** [comment]
I can confirm: it appens on every `poetry self update` ; locked files interrupt the upgrade. I had to reinstall from scratch. Tried both with ps and ps with admin rights

**Comment by jaepil-choi** [comment]
Here's my solution, which I believe is THE EASIEST. 

Just copy paste Poetry installation commend, i.e:
`$ curl -sSL https://raw.githubusercontent.com/python-poetry/poetry/master/get-poetry.py | python`

on your os/linux/bashonwindows terminal, and wait. 

Done.

**Comment by nicolaerario** [comment]
Sorry @jaepil-choi but I can call your solution "re-install it on top"...
I don't think it can be called solution, maybe  workaround.

However, the problem persist on windows (10) and also with new `poetry self update` command; the python version is irrilevant. Wouldn't it be better to adjust the issue title accordingly?

**Comment by jaepil-choi** [comment]
@nicolaerario Yeah, a workaround is a correct term I guess. It does remedy the problem though, because the worst thing that happens when you initially fail update with `poetry self update`, is that `poetry` command just stops working. For those people, this workaround can help. 

Still, I hope this gets fixed in later releases. 
As for issue title, it's not the issue I raised so OP, @ST33LDI9ITAL,  should rename the title.

**Comment by jimbo1qaz** [comment]
I experienced this error upgrading from 1.0.3 to 1.0.5 (latest) via `self upgrade`. Reinstalling 1.0.5 via `curl -sSL https://raw.githubusercontent.com/python-poetry/poetry/master/get-poetry.py | python`, then running `poetry self upgrade 1.0.3`, produced the same error message on `_cmsgpack.cp37-win_amd64.pyd`.

I think I'll try pipx instead of get-poetry.

**Comment by jeriox** [comment]
just ran into the same issue. Any news on this?

**Comment by TBBle** [comment]
Probably the only way Poetry can work around this is by renaming any .pyd file before it is replaced during the update, and not importing any _more_ .pyd modules during the update.

This is the same reason that on Windows, you `python -m pip install --upgrade pip` and not `pip install --upgrade pip`, because the latter locks the _pip.exe_, and you end up with _no_ pip in your venv.

**Comment by CelestialStreamer** [comment]
If I open up PowerShell in administrator mode and run `poetry self update` I get the following error:
```
Traceback (most recent call last):
  File "C:\Users\woo12\.poetry\bin\poetry", line 17, in <module>
    from poetry.console import main
ModuleNotFoundError: No module named 'poetry.console'
```
PowerShell _without_ administrator permissions goes on about those pyd files like you're all talking about.

**Comment by TBBle** [comment]
Running as Administrator is a different issue, it's probably not able to set the path correctly for some reason. Did you install it as Administrator the first time?

**Comment by CelestialStreamer** [comment]
I don't remember how I installed it. I think it might have been with administrator privileges because I must be in administrator mode to run things like `python -m pip install --upgrade pip`. (upgrading pip)
In any case I am not able to upgrade poetry with or without administrator privileges. Due to different errors for each method.

**Comment by TBBle** [comment]
Yeah, I suspect you will have to clean out Poetry entirely (delete `C:\Users\woo12\.poetry`), and reinstall _not_ as administrator, to get it working correctly, as it sounds like part of your install is in Administrator, and part is as a regular user, and something's now conflicting.

I personally use pipx (which avoids a bunch of these issues) so I can't say for sure that there's not other things you'd need to delete. You virtual envs are probably stored in that directory too.

I haven't checked, but perhaps `get-poetry.py` has an "uninstall" option you can run as Administrator to clear out the existing install?

**Comment by Benzidrine** [comment]
I have given up on Poetry due to just experiencing this issue. Best hopes for the future of the project though but I did not find it had the reliability I would want from such a solution.

**Comment by finswimmer** [comment]
The `get-poetry.py` script is deprecated. Please use the `install-poetry.py` script instead and the above mentioned issues should be solved.
